### ISSUE TITLE
Improvements

### ISSUE THREAD

**Issue by osechet** [englercj/tsd-jsdoc#8]
The openlayers library is very well documented and is a good example to test tsd-jsdoc. Running the tool on the library, I get several errors that could help to improve tsd-jsdoc. Here is a list:
- Array of Array: When a type or an argument is defined as `Array<Array<something>>`, only one Array is processed, ending with `Array<something[]>`. The same way, an `Array<Array<Array<something>>>` ends as `Array<Array<somthing[]>>`.
- Inner classes: Inner classes should be generated in a specific module. For example, the `Collection` class defines an `Event` inner class. In Typescript, that `Event` class should be generated in a `Collection` module:

```
class Collection {
...
}

module Collection {
    class Event {
    ...
    }
}
```
- Some jsdoc special characters are not ignored: I think this is an issue with jsdoc but it could be managed easily in tsd-jsdoc. For some complex type definitions, the leading '!' or the trailing '=' are kept. For example, `{ol.geom.Geometry|Object.<string, *>=}` ends in the tsd as `(ol.geom.Geometry|{ [k: string]: any }=)` or `!Array.<string>` ends as `!string[]`. It could be nice to ignore these characters.
- Function prototype: openlayers doc precisely defines the function prototypes. It could be nice to generate equivalent Typescript definition. For example, the `StyleFunction` type is defined as `function((ol.Feature|ol.render.Feature), number): (ol.style.Style|Array.<ol.style.Style>)` but generated as: `() => any`. It should be `(feature: (ol.Feature|ol.render.Feature), resolution: number) => (ol.style.Style|ol.style.Style[])`.
- `{Array.<function(new: ol.format.Feature)>|undefined}` is converted to `((() => any)(new:ol.format.Feature)[] | undefined)`. The first problem is with the argument that is not well positioned. The second problem comes with the name of the argument: 'new'. Arguments name should be checked to ensure they are not keywords. They could be replaced by the first letter of the argument: `(((n: ol.format.Feature) => any)[] | undefined)`.

To reproduce: 
- clone the [openlayers](https://github.com/openlayers/ol3) library
- run `npm install tsd-jsdoc`
- update the `template` line of `config/jsdoc/info/conf.json` with `./node_modules/tsd-jsdoc`
- run `./node_modules/.bin/jsdoc config/jsdoc/api/index.md -c config/jsdoc/api/conf.json --package package.json -d build/tsd`

**Comment by englercj** [comment]
Thanks for the notes, I am working on a few of these now; but some of them are new. I will try to get something going here over the next few weeks but I am pretty busy atm.

**Comment by englercj** [comment]
Getting closer on this, just a few more things to work out and I should be able to release v2 soon. Sorry its been so long :(

**Comment by englercj** [comment]
Got the module declaration stuff working for inner classes and types.

One note about the function stuff is that while I don't support inline function definitions (yet), your use case is supported if you use the JSDoc [`@callback`](http://usejsdoc.org/tags-callback.html) tag to create a named callback, that you then use elsewhere.

You can see an examp[le declaration of a callback [here](https://github.com/englercj/tsd-jsdoc/blob/415a4a66bc9539a4d2cbd006ee89079ef169f798/test/testns/foo.js#L15-L23) and the usage [here](https://github.com/englercj/tsd-jsdoc/blob/415a4a66bc9539a4d2cbd006ee89079ef169f798/test/testns/foo.js#L27).

Thanks for being patient, I'll put out a v2 beta build soon.

**Comment by englercj** [comment]
Just released `v2.0.0-beta.1`, there are still bugs (the issues I opened) but please feel free to try it out and file more bugs!
